### PR TITLE
Next (Review, Feedback)

### PR DESCRIPTION
**Pull request by alekseykulikov**
Overview for the v2.3.0 release. **Please** give us your feedback.
The tradition of [huge rewrites](https://github.com/treosh/lighthouse-ci-action/pull/7) goes on 🚀

**Goals**:
- A better way to communicate results: automatic annotations, slack notification on failure, simple option to collect artifacts for later reuse. #2 #17 #28 #11  
- Integrate with newest features in LHCI and Github Actions: upload artifacts automatically, integrate with annotations, `--uploadUrlMap` to provide [compare report](https://googlechrome.github.io/lighthouse-ci/viewer/) automatically, yaml config
- Improve docs to explain the main benefits and features.

**Expecte workflow**: setup action with budgets and temporary storage, get notification when something breaks ([example](https://github.com/treosh/lighthouse-ci-action/actions/runs/54301567)):
Screenshot 2020-03-12 at 12 42 53

**Changes**:
- `major` Automatic action annotations for failed assertions
- `major` `uploadArtifacts` upload artifacts from the action, instead of extra config, using [@actions/artifact](https://github.com/actions/toolkit/tree/master/packages/artifact)
- Rename `upload.*` to `serverBaseUrl` and `serverToken` with backward compatibility
- Add .yaml config support
- Add support for LHCI's `--uploadUrlMap`, but it does not work
- Refactoring some code to `utils`
- Support for urls in config #29 

**Notes**: 
- This PR is only for review. Due to the massive `node_modules` folder, it's not possible to review the actual changes. So, I reset `node_modules` from the master but kept all the code changes. 
- The working code is ready for testing: https://github.com/treosh/lighthouse-ci-action/tree/next. 
- The README.md and recipes are work in progress.
- The Netlify integration will be closed using community actions #30 #31 (instead of the custom option)

### CODE REVIEW

**Comment by alekseykulikov** [comment]
A lot of time and experiments went to a simple solution for #2 : communicate failed asserts using annotations, attach results to the job, and optionally send slack notification on failure (similar to email, but it's possible to turn email off).

@denar90 and I think this workflow is close to ideal :)
@exterkamp @paulirish let us know what you think.

File: src/utils/lhci-helpers.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+const { groupBy, mapValues, orderBy } = require('lodash')
+const { join } = require('path')
+const fs = require('fs').promises
+const { existsSync } = require('fs')
+
+/**
+ * Get links by url.
```
Comment (exterkamp) [inline src/utils/lhci-helpers.js]: More context please. This gets a `links.json` file via a folder path? And returns the content of that file?

File: src/utils/annotations.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+const core = require('@actions/core')
+const { getLinksByUrl, getAssertionsByUrl } = require('./lhci-helpers')
+
+/**
+ * Set annotations for each failed URL.
+ * 
+ * Example text output (no markdown support for now):
```
Comment (exterkamp) [inline src/utils/annotations.js]:
```suggestion
 * Example text output (no markdown support, https://{link to gh-issue where this is being fixed/issue in this repo}):
```

File: README.md [hunk 3]
```diff
@@ -6,11 +6,11 @@
...
-- ⚙️ Full control over Lighthouse & Lighthouse CI config
-- 🔍 Detailed output for quick debug
-- 💾 Upload data to LHCI server
+- 🎯 Test performance with Lighthouse CI (LHCI) assertions or performance budgets
+- 💾 Upload results to LHCI server or Temporary Public Storage
+- ⚙️ Full control over Lighthouse CI config
+- 😻 Use GitHub to see failed checks
+- 🚀 Fast action initialization (less than 1 second)
 
 <img align="center" width="1046" alt="Lighthouse CI Action" src="https://user-images.githubusercontent.com/158189/68597493-58896f80-049d-11ea-97a2-5c4e7eb4285c.png">
```
Comment (exterkamp) [inline README.md]: Almost want this to be a picture of the new annotations display, or maybe both?

File: src/index.js [hunk 4]
```diff
@@ -1,108 +1,101 @@
...
-
-    if (status !== 0) {
-      // TODO(exterkamp): Output what urls failed and record a nice rich error.
-      core.setFailed(`Assertions have failed.`)
-      // continue
-    }
+    // run lhci with problem matcher
+    // https://github.com/actions/toolkit/blob/master/docs/commands.md#problem-matchers
+    const assertStatus = runChildCommand('assert', assertArgs)
+    isAssertFailed = assertStatus !== 0
```
[the remark this answers is not in the document]
Comment (alekseykulikov) [reply src/index.js]: it is, but `runChildCommand` fallbacks to `-1` (https://github.com/treosh/lighthouse-ci-action/blob/next/src/index.js#L115), so I decided to check explicitly.

File: src/utils/artifacts.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+const artifact = require('@actions/artifact')
+const fs = require('fs').promises
+const { join } = require('path')
+
+/** @param {string} resultsPath */
+exports.uploadArtifacts = async function uploadArtifacts(resultsPath) {
```
[the remark this answers is not in the document]
Comment (alekseykulikov) [reply src/utils/artifacts.js]: There're logs, `uploadArtifact` is explicit about loading data and the process, for example: https://github.com/treosh/lighthouse-ci-action/runs/520099668?check_suite_focus=true
Screenshot 2020-03-19 at 20 12 22

File: README.md (hunk 3 above)
Comment (alekseykulikov) [reply README.md]: yes!

**Comment by alekseykulikov** [comment]
Thank you @exterkamp for a review 👍All the suggestions are in the [`next`](https://github.com/treosh/lighthouse-ci-action/tree/next) branch and going to be merged to master soon.

File: src/index.js [hunk 6]
```diff
@@ -1,108 +1,101 @@
...
-const lhciCliPath = require.resolve('@lhci/cli/src/cli.js')
-const input = require('./input.js')
+const lhciCliPath = require.resolve('@lhci/cli/src/cli')
+const { getInput, hasAssertConfig } = require('./config')
+const { uploadArtifacts } = require('./utils/artifacts')
+const { setFailedAnnotations } = require('./utils/annotations')
+
+/**
+ * Audit urls with Lighthouse CI in 3 stages:
+ * 1. collect (using lhci collect or the custom PSI runner, store results as artifacts)
```
[the remark this answers is not in the document]
Comment (alekseykulikov) [reply src/index.js]: I've been thinking about this issue https://github.com/treosh/lighthouse-ci-action/issues/14 and thought that running on PSI could solve it. But it's out of the scope for now, and I would like to discuss this option in the LHCI repo first.

**Pull request closed without merging**